docs/README.md

```diff
@@ -31,15 +31,41 @@ whichever version of Spark you currently have checked out of revision control.
 The Spark documentation build uses a number of tools to build HTML docs and API docs in Scala, Java,
 Python, R and SQL.
 
-You need to have [Ruby](https://www.ruby-lang.org/en/documentation/installation/) and
-[Python](https://docs.python.org/2/using/unix.html#getting-and-installing-the-latest-version-of-python)
-installed. Also install the following libraries:
+You need to have Ruby 2 and Python 3 installed. A handy way to install and manage various versions of Ruby and Python is with [`rbenv`] and [`pyenv`].
```

Comment:
@nchammas, what about we keep the default way, and simply mention such pyend, virtualenv or conda?

Reply:
We can do that, but I am promoting pyenv and rbenv because they give us a consistent way to install any version of Python and Ruby across Docker and the typical developer environment, without depending on package repositories that typically lag a year or two behind the latest release. pyenv and rbenv also give and simpler and more consistent workflow for getting and using specific versions of Python and Ruby.

With the way things are right now, we don't specify any version of Python or Ruby in the docs README, and in the Dockerfile we use a [4 year old version of Ruby](https://www.ruby-lang.org/en/news/2015/12/25/ruby-2-3-0-released/) and [3 year old version of Python](https://www.python.org/downloads/release/python-360/). Furthermore, to upgrade from these versions to newer ones, developers have to figure out the upgrade path themselves for their local workstation, and for Docker they have to hunt through package repositories that, again, often lag behind what's available.

pyenv and rbenv give us a consistent way to install and manage what versions of Ruby and Python we're using to build docs -- whether on the local machine or in a Docker image -- and they make it dead simple to change the versions we want to use.

Reply:
If the committers are not comfortable with this change, a more conservative change we can make for now is simply to pin the versions of the various Python and Ruby libraries that we install. We can revisit the rbenv/pyenv idea at a later time.

Reply:
I don't know much about these tools, but updating the docs to point to anything newer, if it sounds good to someone looking closely at this part, seems OK.

Reply:
I am not strongly against this change, but not very fond of it as it forces to install unnecessary dependencies. Other people might use conda to manage Python versions as an example. At least I use conda. Given that there are many ways to control this, I am not sure about documenting one way. Why don't we just mention Ruby 2 and Python 3?

Reply:
We can just tell people to install Ruby 2 and Python 3, but without specifying specific versions of them we risk having contributors and committers waste time dealing with library versions that are incompatible with the version of the language/runtime they have installed, or that somehow don't work with our configuration.

It's a small risk, but I figured if we're going to use rbenv and pyenv in building the Docker image, it would make sense to recommend them for the interactive/local workflow as well. The experience is consistent, which is what's nice about using these tools, and they are "lower level" than things like Conda in that they are relatively simple shell scripts that only manage installations.

Are you OK with using rbenv and pyenv in the Docker image?

Reply:
I think docker changes are fine.